### ISSUE TITLE
test/format segfault in split_deepen

### ISSUE THREAD

**Issue by agorrod** [wiredtiger/wiredtiger#1799]
Stack trace:

```
src/btree/bt_split.c:568 (discriminator 2)(__split_deepen)[0x491d4a]
src/btree/bt_split.c:1041(__split_parent)[0x492ccf]
src/btree/bt_split.c:1455(__wt_split_multi)[0x493bc5]
src/evict/evict_page.c:216(__evict_page_dirty_update)[0x42d26f]
src/evict/evict_page.c:118(__wt_evict)[0x42cf1e]
src/evict/evict_lru.c:666(__wt_evict_page)[0x42acd6]
src/evict/evict_lru.c:1394(__wt_evict_lru_page)[0x42c189]
src/evict/evict_lru.c:760 (discriminator 1)(__evict_lru_pages)[0x42aefb]
src/evict/evict_lru.c:396(__evict_worker)[0x42a4ac]
```

Configuration file:

```
############################################
#  RUN PARAMETERS
############################################
abort=0
auto_throttle=1
firstfit=0
bitcnt=4
bloom=1
bloom_bit_count=37
bloom_hash_count=28
bloom_oldest=0
cache=10
checkpoints=1
checksum=uncompressed
chunk_size=4
compaction=0
compression=bzip-raw
data_extend=0
data_source=file
delete_pct=20
dictionary=0
evict_max=3
file_type=row-store
backups=0
huffman_key=0
huffman_value=0
insert_pct=25
internal_key_truncation=1
internal_page_max=11
isolation=read-committed
key_gap=20
key_max=30
key_min=14
leak_memory=0
leaf_page_max=10
logging=0
logging_archive=1
logging_prealloc=0
lsm_worker_threads=3
merge_max=18
mmap=1
ops=100000
prefix_compression=1
prefix_compression_min=6
repeat_data_pct=78
reverse=0
rows=100000
runs=100
split_pct=46
statistics=1
statistics_server=0
threads=11
timer=20
value_max=948
value_min=8
wiredtiger_config=
write_pct=69
############################################
```

http://build.wiredtiger.com:8080/job/wiredtiger-test-format-stress/6433/console

**Comment by agorrod** [comment]
I've attempted to reproduce this without any luck.

I've stared at the code for a while and don't see where it's going wrong. The ref shouldn't be being freed from under us - since we are the ones that free the ```pindex``` just below the loop - so it seems entirely reasonable to use ```WT_INTL_FOREACH_BEGIN_SAFE``` instead of ```WT_INTL_FOREACH_BEGIN``` while walking through this pindex.

It's also my understanding that there can only be a single split at any time - so it should not be possible for another split to free the ref from under us.

I also don't see any recent commits that would introduce a bug here - maybe I'm not looking for the right thing though.

@keithbostic Could you devote some brain cycles to the issue?

**Comment by keithbostic** [comment]
@agorrod, I assume this was develop, but what version of the tree, and what pointer segfaulted?

I think `WT_INTL_FOREACH_BEGIN_SAFE` may not be safe when walking the child page:

- imagine internal page A with internal page children 1-10
- page 2 splits then deepens the tree: creating a new level between page A and pages 1-10
- we are in that loop fixing up child pages (in other words, walking page 3, using page 3's page-index):
- a child of page 3 splits into page 3 and updates/frees page 3's index

In short, we don't have any kind of locking on page 3 or its index, do we?

**Comment by keithbostic** [comment]
@agorrod, given that `WT_ENTER_PAGE_INDEX` only requires a single publish, is it worth having a `_SAFE` version of `WT_INTL_FOREACH_BEGIN`?

**Comment by sueloverso** [comment]
@keithbostic The Jenkins build data says it is from develop, changeset 82b2f1634f1ca68fbd9701569b3adaf2c6f3d6d5.  I've unpacked the tar on tinderbox in stress.6433.  Here's the stack:
```
#0  0x0000000000491d4a in __split_deepen (session=0x1cae7c0, parent=0x1caba20, 
    children=724) at ../src/btree/bt_split.c:568
#1  0x0000000000492ccf in __split_parent (session=0x1cae7c0, ref=0x0, 
    ref_new=0x7fe7c02dd590, new_entries=2, parent_incr=58100, exclusive=0)
    at ../src/btree/bt_split.c:1041
#2  0x0000000000493bc5 in __wt_split_multi (session=0x1cae7c0, 
    ref=0x7fe7b8058510, exclusive=0) at ../src/btree/bt_split.c:1455
#3  0x000000000042d26f in __evict_page_dirty_update (session=0x1cae7c0, 
    ref=0x7fe7b8058510, exclusive=0) at ../src/evict/evict_page.c:216
#4  0x000000000042cf1e in __wt_evict (session=0x1cae7c0, ref=0x7fe7b8058510, 
    exclusive=0) at ../src/evict/evict_page.c:118
#5  0x000000000042acd6 in __wt_evict_page (session=0x1cae7c0, 
    ref=0x7fe7b8058510) at ../src/evict/evict_lru.c:666
#6  0x000000000042c189 in __wt_evict_lru_page (session=0x1cae7c0, is_server=0)
    at ../src/evict/evict_lru.c:1394
#7  0x000000000042aefb in __evict_lru_pages (session=0x1cae7c0, is_server=0)
    at ../src/evict/evict_lru.c:760
#8  0x000000000042a4ac in __evict_worker (arg=0x1c95300)
    at ../src/evict/evict_lru.c:396
#9  0x00000037ce207ee5 in start_thread (arg=0x7fe836ffd700)
    at pthread_create.c:309
```
The `child_ref` pointer is bad:
```
568				if (child_ref->home == parent) {
(gdb) p child_ref
$1 = (WT_REF *) 0x16f0
(gdb) p parent
$2 = (WT_PAGE *) 0x1caba20
(gdb) p child
$3 = (WT_PAGE *) 0x7fe7c0192a00
(gdb) p *child
$4 = {u = {intl = {recno = 0, parent_ref = 0x7fe7c02b1fe0, 
      __index = 0x7fe7fc643870}, row = {ins = 0x0, upd = 0x7fe7c02b1fe0, 
      d = 0x7fe7fc643870, entries = 0}, col_fix = {recno = 0, 
      bitf = 0x7fe7c02b1fe0 "", entries = 4234426480}, col_var = {recno = 0, 
      d = 0x7fe7c02b1fe0, repeats = 0x7fe7fc643870, nrepeats = 0, 
      entries = 0}}, type = 6 '\006', flags_atomic = 128 '\200', 
  read_gen = 7366, memory_footprint = 148495, dsk = 0x0, 
  modify = 0x7fe7c0194170}
```

**Comment by keithbostic** [comment]
Obviously, `child_ref` is borked. Hmmm, 0x6f is an "o", any bytes around `child_ref` that might indicate it's part of memory that's been freed and is now being re-used?

It also might be worth digging to see if `child` has recently been split, although I can't offhand think of a way to figure that out.

**Comment by sueloverso** [comment]
It sort of looks like the entries count is off by one and `__pindex` looks overwritten or freed. 
```
(gdb) p __refp
$3 = (WT_REF **) 0x7fe7c0194138
(gdb) p *__refp
$4 = (WT_REF *) 0x34
(gdb) p *__pindex
$5 = {entries = 3222880816, index = 0x7fe7c0000078}
(gdb) p __entries
$6 = 1
```
Looking at the memory before/around `__refp` we see addresses until we get to the failing one at 
0x7fe7c0194130:
```
(gdb) x/100x 0x7fe7c0194000
0x7fe7c0194000:	0xb8286e00	0x00007fe7	0xb805c730	0x00007fe7
0x7fe7c0194010:	0xb806c5a0	0x00007fe7	0xb4220b90	0x00007fe7
0x7fe7c0194020:	0xc019d150	0x00007fe7	0x2805eb60	0x00007fe8
0x7fe7c0194030:	0xb808e030	0x00007fe7	0xb82536f0	0x00007fe7
0x7fe7c0194040:	0xc0119010	0x00007fe7	0xc010cd90	0x00007fe7
0x7fe7c0194050:	0x0c0c4db0	0x00007fe8	0x2c06e210	0x00007fe8
0x7fe7c0194060:	0x3c0226c0	0x00007fe8	0xb808fcd0	0x00007fe7
0x7fe7c0194070:	0xb40be750	0x00007fe7	0x3c00f350	0x00007fe8
0x7fe7c0194080:	0x24041e90	0x00007fe8	0xb808de90	0x00007fe7
0x7fe7c0194090:	0xb8285ee0	0x00007fe7	0x140a8a40	0x00007fe8
0x7fe7c01940a0:	0xb80b76a0	0x00007fe7	0xac047640	0x00007fe7
0x7fe7c01940b0:	0x3c248010	0x00007fe8	0x3c24a140	0x00007fe8
0x7fe7c01940c0:	0x3c2382e0	0x00007fe8	0x0c0ec160	0x00007fe8
0x7fe7c01940d0:	0xac18ed90	0x00007fe7	0x08047dd0	0x00007fe8
0x7fe7c01940e0:	0x2c088300	0x00007fe8	0x280c9500	0x00007fe8
0x7fe7c01940f0:	0x280dabc0	0x00007fe8	0x3c1f6db0	0x00007fe8
0x7fe7c0194100:	0xfc041c30	0x00007fe7	0xac0a9ab0	0x00007fe7
0x7fe7c0194110:	0x140b9600	0x00007fe8	0x3c239250	0x00007fe8
0x7fe7c0194120:	0x080424a0	0x00007fe8	0x08259f40	0x00007fe8
0x7fe7c0194130:	0x000016f0	0x00000000	0x00000034	0x00000000
```

**Comment by sueloverso** [comment]
The absurd entries value shown in `__pindex` is really the bottom 32 bits of an address:
```
(gdb) p __pindex
$7 = (WT_PAGE_INDEX *) 0x7fe7c0192a50
(gdb) x/40x 0x7fe7c0192a00
0x7fe7c0192a00:	0x00000000	0x00000000	0xc02b1fe0	0x00007fe7
0x7fe7c0192a10:	0xfc643870	0x00007fe7	0x00000000	0x00000000
0x7fe7c0192a20:	0x00008006	0x00000000	0x00001cc6	0x00000000
0x7fe7c0192a30:	0x0002440f	0x00000000	0x00000000	0x00000000
0x7fe7c0192a40:	0xc0194170	0x00007fe7	0x000016f1	0x00000000
0x7fe7c0192a50:	0xc0194230	0x00007fe7	0xc0000078	0x00007fe7
0x7fe7c0192a60:	0x00000000	0x00000000	0x00000000	0x00000000
0x7fe7c0192a70:	0x080c1b10	0x00007fe8	0xd0069f50	0x00007fe7
0x7fe7c0192a80:	0x0c0778a0	0x00007fe8	0x0c0c9620	0x00007fe8
0x7fe7c0192a90:	0xb8090700	0x00007fe7	0x3c01f2a0	0x00007fe8
```
If I look a bit more at the memory around `__refp` I see this:
```
0x7fe7c0194130:	0x000016f0	0x00000000	0x00000034	0x00000000
0x7fe7c0194140:	0x0000001b	0x00000000	0x30303030	0x36313830
0x7fe7c0194150:	0x302e3838	0x706f2f32	0x74737271	0x78777675
0x7fe7c0194160:	0x00617a79	0x00000000	0x00000075	0x00000000
0x7fe7c0194170:	0x00000000	0x00000000	0x00000000	0x00000000
0x7fe7c0194180:	0x0000af15	0x00000000	0x00000000	0x00000000
(gdb) x/s 0x7fe7c0194148
0x7fe7c0194148:	"0000081688.02/opqrstuvwxyza"
```

**Comment by agorrod** [comment]
> given that WT_ENTER_PAGE_INDEX only requires a single publish, is it worth having a _SAFE version of WT_INTL_FOREACH_BEGIN?

The primary difference between those macros is the assertion the ```split_gen``` is set. In the case where we are doing a split we don't necessarily have a split generation (so safe_free can theoretically go ahead and free the memory).

I didn't think we allowed splitting at multiple levels in a tree concurrently (hence I ruled out your scenario above). I'm not sure why I thought that. If it's possible that we are splitting into a child while walking it for a split into the parent (yuck), we need to have a split_gen set to avoid exactly the problem seen here.
